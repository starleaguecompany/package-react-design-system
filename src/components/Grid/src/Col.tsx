import * as React from 'react';
import { useStyles } from '@starleaguecompany/package-react-utils';

import { ColProps } from '../types/Col.types';
import RowContext from './RowContext';
import styles from '../styles/Col.module.less';

/**
 * @description Column element for Grid system.
 *
 * @component
 * @example
 * ```jsx
 * <Col span={3}>
 *   Column
 * </Button>
 * ```
 */
const Col = React.forwardRef<HTMLDivElement, ColProps>((props, ref) => {
  const { className, span, order, breakpoints, style, children, ...restProps } = props;
  const cx = useStyles(styles);
  // const [match] = useMediaQuery()

  // const breakPointProps = React.useMemo(() => {
  //   if (breakpoints && match) {
  //     // @ts-ignore
  //     return breakpoints[match] || {}
  //   }
  //
  //   return {}
  // }, [match, breakpoints])

  const classNames = cx(className, 'container', {
    // [`span-${breakPointProps.span || span}`]: breakPointProps.span || span,
    [`span-${span}`]: span,
  });
  const { gutter } = React.useContext(RowContext);
  const mergedStyle: React.CSSProperties = {
    ...style,
    // order: breakPointProps.order || order || undefined,
    order: order || undefined,
    // ...breakPointProps.style,
  };

  // Horizontal gutter use padding
  if (gutter && gutter[0] > 0) {
    const horizontalGutter = gutter[0] / 2;
    mergedStyle.paddingLeft = horizontalGutter;
    mergedStyle.paddingRight = horizontalGutter;
  }

  // Vertical gutter use padding when gap not support
  if (gutter && gutter[1] > 0) {
    const verticalGutter = gutter[1] / 2;
    mergedStyle.paddingTop = verticalGutter;
    mergedStyle.paddingBottom = verticalGutter;
  }

  return (
    <div data-qa="Col" ref={ref} className={classNames} style={{ ...mergedStyle }} {...restProps}>
      {children}
    </div>
  );
});

Col.defaultProps = {};

export default Col;
