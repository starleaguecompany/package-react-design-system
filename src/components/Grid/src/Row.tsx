import * as React from 'react'
import { useStyles } from '@starleaguecompany/package-react-utils'

import { RowProps } from '../types/Row.types'
import RowContext from './RowContext'
import styles from '../styles/Row.module.less'

/**
 * @description Container element for Grid system.
 *
 * @component
 * @example
 * ```jsx
 * <Row gutter={[20, 20]} wrap>
 *   ...columns
 * </Button>
 * ```
 */
const Row = React.forwardRef<HTMLDivElement, RowProps>((props, ref) => {
  const { className, gutter, align = 'top', justify, wrap: wrap, style, children, ...restProps } = props
  const cx = useStyles(styles)

  const classNames = cx(className, 'container', {
    wrap: wrap,
    [`align-${align}`]: align,
    [`justify-${justify}`]: justify,
  })

  const gutters = React.useMemo(() => {
    const results: [number, number] = [0, 0]
    const normalizedGutter = Array.isArray(gutter) ? gutter : [gutter, 0]

    normalizedGutter.forEach((g, index) => {
      results[index] = g || 0
    })

    return results
  }, [gutter])

  const rowStyle: React.CSSProperties = { ...style }
  const horizontalGutter = gutters[0] > 0 ? gutters[0] / -2 : undefined

  if (horizontalGutter) {
    rowStyle.marginLeft = horizontalGutter
    rowStyle.marginRight = horizontalGutter
  }

  const rowContext = React.useMemo(() => ({ gutter: gutters, wrap }), [gutters, wrap])

  return (
    <RowContext.Provider value={rowContext}>
      <div data-qa="Row" ref={ref} className={classNames} style={{ ...rowStyle }} {...restProps}>
        {children}
      </div>
    </RowContext.Provider>
  )
})

Row.defaultProps = {
  gutter: 0,
  align: 'top',
  justify: 'start',
  wrap: true,
}

export default Row
