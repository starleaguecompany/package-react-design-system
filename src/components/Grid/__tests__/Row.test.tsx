import * as React from 'react';
import { render, screen } from '@testing-library/react';

import Row from '../src/Row';
import Col from '../src/Col';

describe('Row', () => {
  test('should render correctly with no props', () => {
    render(<Row />);

    expect(screen.getByTestId('Row')).toMatchSnapshot();
    expect(screen.getByTestId('Row')).toHaveClass('container wrap align-top justify-start', { exact: true });
  });

  test('should render correctly with attributes', () => {
    render(<Row id="test-id" className="test-class" />);

    expect(screen.getByTestId('Row')).toMatchSnapshot();
    expect(screen.getByTestId('Row')).toHaveClass('test-class');
    expect(screen.getByTestId('Row')).toHaveAttribute('id', 'test-id');
  });

  test('should render correctly with text', () => {
    render(<Row>Testing row</Row>);

    expect(screen.getByTestId('Row')).toMatchSnapshot();
    expect(screen.getByTestId('Row')).toHaveTextContent('Testing row');
    expect(screen.getByText('Testing row')).toBeVisible();
  });

  test('should render correctly with gutter prop', () => {
    render(
      <Row gutter={[10, 30]}>
        <Col>Testing value 1</Col>
        <Col>Testing value 2</Col>
      </Row>
    );

    expect(screen.getByTestId('Row')).toMatchSnapshot();
    expect(screen.getByTestId('Row')).toHaveStyle('margin-left: -5px; margin-right: -5px;');
    expect(screen.getByText('Testing value 1')).toHaveStyle('padding: 15px 5px 15px 5px;');
    expect(screen.getByText('Testing value 2')).toHaveStyle('padding: 15px 5px 15px 5px;');
  });

  test('should render correctly with align prop', () => {
    render(<Row align="bottom">Testing row</Row>);

    expect(screen.getByTestId('Row')).toMatchSnapshot();
    expect(screen.getByTestId('Row')).toHaveClass('align-bottom');
  });

  test('should render correctly with justify prop', () => {
    render(<Row justify="center">Testing row</Row>);

    expect(screen.getByTestId('Row')).toMatchSnapshot();
    expect(screen.getByTestId('Row')).toHaveClass('justify-center');
  });

  test('should render correctly with wrap prop', () => {
    render(<Row wrap={false}>Testing row</Row>);

    expect(screen.getByTestId('Row')).toMatchSnapshot();
    expect(screen.getByTestId('Row')).not.toHaveClass('wrap');
  });
});
