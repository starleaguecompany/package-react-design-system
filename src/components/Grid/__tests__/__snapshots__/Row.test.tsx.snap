// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row should render correctly with align prop 1`] = `
<div
  class="container wrap align-bottom justify-start"
  data-qa="Row"
>
  Testing row
</div>
`;

exports[`Row should render correctly with attributes 1`] = `
<div
  class="test-class container wrap align-top justify-start"
  data-qa="Row"
  id="test-id"
/>
`;

exports[`Row should render correctly with gutter prop 1`] = `
<div
  class="container wrap align-top justify-start"
  data-qa="Row"
  style="margin-left: -5px; margin-right: -5px;"
>
  <div
    class="container"
    data-qa="Col"
    style="padding: 15px 5px 15px 5px;"
  >
    Testing value 1
  </div>
  <div
    class="container"
    data-qa="Col"
    style="padding: 15px 5px 15px 5px;"
  >
    Testing value 2
  </div>
</div>
`;

exports[`Row should render correctly with justify prop 1`] = `
<div
  class="container wrap align-top justify-center"
  data-qa="Row"
>
  Testing row
</div>
`;

exports[`Row should render correctly with no props 1`] = `
<div
  class="container wrap align-top justify-start"
  data-qa="Row"
/>
`;

exports[`Row should render correctly with text 1`] = `
<div
  class="container wrap align-top justify-start"
  data-qa="Row"
>
  Testing row
</div>
`;

exports[`Row should render correctly with wrap prop 1`] = `
<div
  class="container align-top justify-start"
  data-qa="Row"
>
  Testing row
</div>
`;
